File: examples/showcase/superset_fronting/custom-code-bundles/superset_fronting_java/src/main/java/{{ANIP_JAVA_PACKAGE_PATH}}/BackendAdapter.java
package {{ANIP_JAVA_PACKAGE_NAME}};

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import dev.anip.service.InvocationContext;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

@FunctionalInterface
public interface BackendAdapter {

    Map<String, Object> execute(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> adapterInput, InvocationContext context);

    static BackendAdapter defaultAdapter() {
        return new SupersetBackendAdapter();
    }
}

final class SupersetBackendAdapter implements BackendAdapter {
    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final HttpClient HTTP = HttpClient.newHttpClient();

    @Override
    public Map<String, Object> execute(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, InvocationContext context) {
        List<String> unresolved = stringList(plan.getOrDefault("unresolved_required_backend_inputs", List.of()));
        if (!unresolved.isEmpty()) return result(capability, plan, "backend_input_incomplete", record("unresolved_required_backend_inputs", unresolved));
        String token = accessToken();
        if (token.isBlank()) return result(capability, plan, "backend_error", record("superset_error", record("error", "missing_superset_credentials")));
        return switch (text(capability.get("capability_id"))) {
            case "superset.analytics.discover_context" -> discoverContext(capability, plan, params, token);
            case "superset.analytics.answer_question" -> answerQuestion(capability, plan, params);
            case "superset.chart.preview.create" -> chartPreview(capability, plan, params);
            case "superset.chart.publish.request" -> chartPublishRequest(capability, plan, params, context);
            case "superset.dashboard.draft.prepare" -> dashboardDraft(capability, plan, params);
            case "superset.dataset.draft.prepare" -> datasetDraft(capability, plan, params, context);
            default -> result(capability, plan, "backend_execution_stub", record("note", "No Superset custom handler is registered for this capability."));
        };
    }

    private static Map<String, Object> discoverContext(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, String token) {
        String workspaceScope = text(params.get("workspace_scope"));
        if (!scopeAllowed(workspaceScope)) return restricted(capability, plan, "Workspace scope is outside the configured ANIP policy.");
        String query = text(params.get("query")).toLowerCase();
        int limit = boundedLimit(params.get("limit"), 20, 50);
        String assetType = text(params.get("asset_type"));
        List<String[]> endpoints = List.of(new String[]{"dataset", "/api/v1/dataset/"}, new String[]{"chart", "/api/v1/chart/"}, new String[]{"dashboard", "/api/v1/dashboard/"});
        List<Map<String, Object>> items = new ArrayList<>();
        for (String[] endpoint : endpoints) {
            String kind = endpoint[0];
            if (!assetType.isBlank() && !assetType.equals(kind)) continue;
            Map<String, Object> payload = requestJson("GET", endpoint[1] + "?page_size=" + limit, token, null);
            if (payload.containsKey("error")) return result(capability, plan, "backend_error", record("superset_error", payload));
            for (Map<String, Object> item : listResult(payload)) {
                String title = firstNonBlank(item.get("table_name"), item.get("slice_name"), item.get("dashboard_title"), item.get("name"), item.get("id"));
                if (!query.isBlank() && !title.toLowerCase().contains(query)) continue;
                items.add(record("asset_type", kind, "id", item.get("id"), "title", title, "url", item.get("url")));
                if (items.size() >= limit) break;
            }
            if (items.size() >= limit) break;
        }
        return result(capability, plan, "completed", record("result", record("workspace_scope", workspaceScope, "items", items, "count", items.size())));
    }

    private static Map<String, Object> answerQuestion(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params) {
        String datasetRef = text(params.get("dataset_ref"));
        if (!datasetAllowed(datasetRef)) return restricted(capability, plan, "Dataset is outside the configured ANIP policy.");
        return result(capability, plan, "completed", record("mutation_performed", false, "result", record("question", params.get("question"), "dataset_ref", datasetRef, "metric", params.get("metric"), "dimension", params.get("dimension"), "time_window", params.get("time_window"), "answer", "Governed analytics answer placeholder. The service owns SQL generation and execution policy.", "raw_sql_disclosed", false)));
    }

    private static Map<String, Object> chartPreview(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params) {
        String datasetRef = text(params.get("dataset_ref"));
        if (!datasetAllowed(datasetRef)) return restricted(capability, plan, "Dataset is outside the configured ANIP policy.");
        Map<String, Object> body = record("dataset_ref", datasetRef, "metric", params.get("metric"), "dimension", params.get("dimension"), "visualization_type", params.get("visualization_type"), "title", firstNonBlank(params.get("title"), text(params.get("metric")) + " by " + firstNonBlank(params.get("dimension"), "time")), "save_chart", false);
        return writePreview(capability, plan, "chart.preview", body, record("dataset_ref", datasetRef));
    }

    private static Map<String, Object> chartPublishRequest(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, InvocationContext context) {
        Map<String, Object> preview = writePreview(capability, plan, "chart.publish", record("chart_preview_ref", params.get("chart_preview_ref"), "dashboard_scope", params.get("dashboard_scope"), "reason", params.get("reason"), "title", params.get("title")), record("dashboard_scope", params.get("dashboard_scope")));
        if ("true".equals(System.getenv("ANIP_SUPERSET_ALLOW_MUTATION")) && context != null && !text(context.getApprovalGrant()).isBlank()) {
            preview.put("execution_status", "completed");
            preview.put("approval_required", false);
            preview.put("mutation_performed", false);
            preview.put("note", "Approved publish request recorded. Concrete chart save is intentionally left to deployment-specific Superset adapter code.");
        }
        return preview;
    }

    private static Map<String, Object> dashboardDraft(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params) {
        return writePreview(capability, plan, "dashboard.draft", record("dashboard_scope", params.get("dashboard_scope"), "objective", params.get("objective"), "chart_refs", params.get("chart_refs"), "layout_hint", params.get("layout_hint"), "audience", params.get("audience")), record("dashboard_scope", params.get("dashboard_scope")));
    }

    private static Map<String, Object> datasetDraft(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, InvocationContext context) {
        Map<String, Object> preview = writePreview(capability, plan, "dataset.draft", record("database_ref", params.get("database_ref"), "dataset_purpose", params.get("dataset_purpose"), "query_intent", params.get("query_intent"), "source_tables", params.get("source_tables"), "metrics", params.get("metrics"), "raw_sql_accepted", false), record("database_ref", params.get("database_ref")));
        if ("true".equals(System.getenv("ANIP_SUPERSET_ALLOW_MUTATION")) && context != null && !text(context.getApprovalGrant()).isBlank()) {
            preview.put("execution_status", "completed");
            preview.put("approval_required", false);
            preview.put("mutation_performed", false);
            preview.put("note", "Approved dataset draft recorded. Raw SQL generation remains deployment-owned.");
        }
        return preview;
    }

    private static Map<String, Object> writePreview(Map<String, Object> capability, Map<String, Object> plan, String action, Map<String, Object> body, Map<String, Object> metadata) {
        return result(capability, plan, "prepared", record("approval_required", true, "mutation_performed", false, "superset_action", action, "superset_metadata", metadata, "superset_request", record("operation", action, "body", body), "note", "Prepared a governed Superset analytics request. No Superset mutation was performed."));
    }

    private static String accessToken() {
        String direct = text(System.getenv("SUPERSET_ACCESS_TOKEN"));
        if (!direct.isBlank()) return direct;
        String username = text(System.getenv("SUPERSET_USERNAME"));
        String password = text(System.getenv("SUPERSET_PASSWORD"));
        if (username.isBlank() || password.isBlank()) return "";
        Map<String, Object> payload = requestJson("POST", "/api/v1/security/login", "", record("username", username, "password", password, "provider", firstNonBlank(System.getenv("SUPERSET_AUTH_PROVIDER"), "db"), "refresh", true));
        return text(payload.get("access_token"));
    }

    private static Map<String, Object> requestJson(String method, String path, String token, Map<String, Object> body) {
        try {
            HttpRequest.Builder builder = HttpRequest.newBuilder(URI.create(firstNonBlank(System.getenv("SUPERSET_BASE_URL"), "http://127.0.0.1:18088").replaceAll("/$", "") + path))
                .header("Accept", "application/json")
                .header("Content-Type", "application/json")
                .header("User-Agent", "anip-superset-fronting-showcase");
            if (!token.isBlank()) builder.header("Authorization", "Bearer " + token);
            HttpRequest request = builder.method(method, body == null ? HttpRequest.BodyPublishers.noBody() : HttpRequest.BodyPublishers.ofString(MAPPER.writeValueAsString(body))).build();
            HttpResponse<String> response = HTTP.send(request, HttpResponse.BodyHandlers.ofString());
            Map<String, Object> payload = response.body().isBlank() ? new LinkedHashMap<>() : MAPPER.readValue(response.body(), new TypeReference<>() {});
            if (response.statusCode() < 200 || response.statusCode() >= 300) return record("error", "superset_http_error", "status", response.statusCode(), "detail", payload);
            return payload;
        } catch (Exception exc) {
            return record("error", "superset_connection_error", "detail", exc.toString());
        }
    }

    private static Map<String, Object> result(Map<String, Object> capability, Map<String, Object> plan, String status, Map<String, Object> extra) {
        Map<String, Object> result = record("execution_status", status, "capability_id", capability.get("capability_id"), "selected_backend", plan.get("selected_binding"), "semantic_input", plan.get("semantic_input"), "backend_input_contract", plan.get("backend_input_contract"));
        result.putAll(extra);
        return result;
    }

    private static Map<String, Object> restricted(Map<String, Object> capability, Map<String, Object> plan, String reason) {
        return result(capability, plan, "restricted", record("reason", reason));
    }

    private static boolean scopeAllowed(String scope) {
        String key = scope.toLowerCase().trim();
        List<String> blocked = csvEnv("ANIP_SUPERSET_BLOCKED_WORKSPACES");
        List<String> allowed = csvEnv("ANIP_SUPERSET_ALLOWED_WORKSPACES");
        return !blocked.contains(key) && (allowed.isEmpty() || allowed.contains(key));
    }

    private static boolean datasetAllowed(String datasetRef) {
        List<String> allowed = csvEnv("ANIP_SUPERSET_ALLOWED_DATASETS");
        return allowed.isEmpty() || allowed.contains(datasetRef.toLowerCase().trim());
    }

    @SuppressWarnings("unchecked")
    private static List<Map<String, Object>> listResult(Map<String, Object> payload) {
        if (!(payload.get("result") instanceof Map<?, ?> raw)) return List.of();
        Object data = raw.get("data");
        if (data instanceof List<?> list) return (List<Map<String, Object>>) (List<?>) list;
        return List.of();
    }

    private static int boundedLimit(Object value, int defaultValue, int maximum) {
        try {
            int parsed = Integer.parseInt(text(value));
            return Math.max(1, Math.min(parsed, maximum));
        } catch (Exception ignored) {
            return defaultValue;
        }
    }

    private static Map<String, Object> record(Object... pairs) {
        Map<String, Object> result = new LinkedHashMap<>();
        for (int index = 0; index + 1 < pairs.length; index += 2) result.put(String.valueOf(pairs[index]), pairs[index + 1]);
        return result;
    }

    private static String firstNonBlank(Object... values) {
        for (Object value : values) {
            String text = text(value);
            if (!text.isBlank()) return text;
        }
        return "";
    }

    private static String text(Object value) {
        return value == null ? "" : String.valueOf(value).trim();
    }

    private static List<String> stringList(Object value) {
        if (value instanceof List<?> list) return list.stream().map(SupersetBackendAdapter::text).filter(item -> !item.isBlank()).distinct().toList();
        if (value instanceof String text && !text.isBlank()) return List.of(text.split(",")).stream().map(String::trim).filter(item -> !item.isBlank()).distinct().toList();
        return List.of();
    }

    private static List<String> csvEnv(String name) {
        return stringList(System.getenv(name)).stream().map(String::toLowerCase).toList();
    }
}

File: examples/showcase/superset_fronting/custom-code-bundles/superset_fronting_csharp/BackendAdapter.cs
using Anip.Service;
using System.Text;
using System.Text.Json;

namespace {{ANIP_CSHARP_ROOT_NAMESPACE}};

public delegate Dictionary<string, object?> BackendAdapterHandler(
    Dictionary<string, object?> capability,
    Dictionary<string, object?> plan,
    Dictionary<string, object?> adapterInput,
    InvocationContext context);

public static class BackendAdapter
{
    public static BackendAdapterHandler Default => Execute;
    private static readonly HttpClient Http = new();

    private static Dictionary<string, object?> Execute(
        Dictionary<string, object?> capability,
        Dictionary<string, object?> plan,
        Dictionary<string, object?> adapterInput,
        InvocationContext context)
    {
        var unresolved = GeneratedCapabilities.StringList(plan.GetValueOrDefault("unresolved_required_backend_inputs"));
        if (unresolved.Count > 0)
        {
            return Result(capability, plan, "backend_input_incomplete", new() { ["unresolved_required_backend_inputs"] = unresolved });
        }

        var token = AccessToken();
        if (string.IsNullOrWhiteSpace(token))
        {
            return Result(capability, plan, "backend_error", new() { ["superset_error"] = new Dictionary<string, object?> { ["error"] = "missing_superset_credentials" } });
        }

        return Text(capability.GetValueOrDefault("capability_id")) switch
        {
            "superset.analytics.discover_context" => DiscoverContext(capability, plan, adapterInput, token),
            "superset.analytics.answer_question" => AnswerQuestion(capability, plan, adapterInput),
            "superset.chart.preview.create" => ChartPreview(capability, plan, adapterInput),
            "superset.chart.publish.request" => ChartPublishRequest(capability, plan, adapterInput, context),
            "superset.dashboard.draft.prepare" => DashboardDraft(capability, plan, adapterInput),
            "superset.dataset.draft.prepare" => DatasetDraft(capability, plan, adapterInput, context),
            _ => Result(capability, plan, "backend_execution_stub", new() { ["note"] = "No Superset custom handler is registered for this capability." }),
        };
    }

    private static Dictionary<string, object?> DiscoverContext(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> parameters, string token)
    {
        var workspaceScope = Text(parameters.GetValueOrDefault("workspace_scope"));
        if (!ScopeAllowed(workspaceScope)) return Restricted(capability, plan, "Workspace scope is outside the configured ANIP policy.");
        var query = Text(parameters.GetValueOrDefault("query")).ToLowerInvariant();
        var limit = BoundedLimit(parameters.GetValueOrDefault("limit"), 20, 50);
        var assetType = Text(parameters.GetValueOrDefault("asset_type"));
        var endpoints = new[] { ("dataset", "/api/v1/dataset/"), ("chart", "/api/v1/chart/"), ("dashboard", "/api/v1/dashboard/") };
        var items = new List<Dictionary<string, object?>>();
        foreach (var (kind, endpoint) in endpoints)
        {
            if (!string.IsNullOrWhiteSpace(assetType) && assetType != kind) continue;
            var payload = RequestJson("GET", $"{endpoint}?page_size={limit}", token, null);
            if (payload.ContainsKey("error")) return Result(capability, plan, "backend_error", new() { ["superset_error"] = payload });
            foreach (var item in ListResult(payload))
            {
                var title = FirstNonEmpty(Get(item, "table_name"), Get(item, "slice_name"), Get(item, "dashboard_title"), Get(item, "name"), Get(item, "id"));
                if (!string.IsNullOrWhiteSpace(query) && !title.ToLowerInvariant().Contains(query)) continue;
                items.Add(new() { ["asset_type"] = kind, ["id"] = Get(item, "id"), ["title"] = title, ["url"] = Get(item, "url") });
                if (items.Count >= limit) break;
            }
            if (items.Count >= limit) break;
        }
        return Result(capability, plan, "completed", new() { ["result"] = new Dictionary<string, object?> { ["workspace_scope"] = workspaceScope, ["items"] = items, ["count"] = items.Count } });
    }

    private static Dictionary<string, object?> AnswerQuestion(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> parameters)
    {
        var datasetRef = Text(parameters.GetValueOrDefault("dataset_ref"));
        if (!DatasetAllowed(datasetRef)) return Restricted(capability, plan, "Dataset is outside the configured ANIP policy.");
        return Result(capability, plan, "completed", new()
        {
            ["mutation_performed"] = false,
            ["result"] = new Dictionary<string, object?>
            {
                ["question"] = parameters.GetValueOrDefault("question"),
                ["dataset_ref"] = datasetRef,
                ["metric"] = parameters.GetValueOrDefault("metric"),
                ["dimension"] = parameters.GetValueOrDefault("dimension"),
                ["time_window"] = parameters.GetValueOrDefault("time_window"),
                ["answer"] = "Governed analytics answer placeholder. The service owns SQL generation and execution policy.",
                ["raw_sql_disclosed"] = false,
            },
        });
    }

    private static Dictionary<string, object?> ChartPreview(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> parameters)
    {
        var datasetRef = Text(parameters.GetValueOrDefault("dataset_ref"));
        if (!DatasetAllowed(datasetRef)) return Restricted(capability, plan, "Dataset is outside the configured ANIP policy.");
        var body = new Dictionary<string, object?>
        {
            ["dataset_ref"] = datasetRef,
            ["metric"] = parameters.GetValueOrDefault("metric"),
            ["dimension"] = parameters.GetValueOrDefault("dimension"),
            ["visualization_type"] = parameters.GetValueOrDefault("visualization_type"),
            ["title"] = FirstNonEmpty(parameters.GetValueOrDefault("title"), $"{Text(parameters.GetValueOrDefault("metric"))} by {FirstNonEmpty(parameters.GetValueOrDefault("dimension"), "time")}"),
            ["save_chart"] = false,
        };
        return WritePreview(capability, plan, "chart.preview", body, new() { ["dataset_ref"] = datasetRef });
    }

    private static Dictionary<string, object?> ChartPublishRequest(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> parameters, InvocationContext context)
    {
        var preview = WritePreview(capability, plan, "chart.publish", new() { ["chart_preview_ref"] = parameters.GetValueOrDefault("chart_preview_ref"), ["dashboard_scope"] = parameters.GetValueOrDefault("dashboard_scope"), ["reason"] = parameters.GetValueOrDefault("reason"), ["title"] = parameters.GetValueOrDefault("title") }, new() { ["dashboard_scope"] = parameters.GetValueOrDefault("dashboard_scope") });
        if (Environment.GetEnvironmentVariable("ANIP_SUPERSET_ALLOW_MUTATION") == "true" && !string.IsNullOrWhiteSpace(context?.ApprovalGrant))
        {
            preview["execution_status"] = "completed";
            preview["approval_required"] = false;
            preview["mutation_performed"] = false;
            preview["note"] = "Approved publish request recorded. Concrete chart save is intentionally left to deployment-specific Superset adapter code.";
        }
        return preview;
    }

    private static Dictionary<string, object?> DashboardDraft(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> parameters) =>
        WritePreview(capability, plan, "dashboard.draft", new() { ["dashboard_scope"] = parameters.GetValueOrDefault("dashboard_scope"), ["objective"] = parameters.GetValueOrDefault("objective"), ["chart_refs"] = parameters.GetValueOrDefault("chart_refs"), ["layout_hint"] = parameters.GetValueOrDefault("layout_hint"), ["audience"] = parameters.GetValueOrDefault("audience") }, new() { ["dashboard_scope"] = parameters.GetValueOrDefault("dashboard_scope") });

    private static Dictionary<string, object?> DatasetDraft(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> parameters, InvocationContext context)
    {
        var preview = WritePreview(capability, plan, "dataset.draft", new() { ["database_ref"] = parameters.GetValueOrDefault("database_ref"), ["dataset_purpose"] = parameters.GetValueOrDefault("dataset_purpose"), ["query_intent"] = parameters.GetValueOrDefault("query_intent"), ["source_tables"] = parameters.GetValueOrDefault("source_tables"), ["metrics"] = parameters.GetValueOrDefault("metrics"), ["raw_sql_accepted"] = false }, new() { ["database_ref"] = parameters.GetValueOrDefault("database_ref") });
        if (Environment.GetEnvironmentVariable("ANIP_SUPERSET_ALLOW_MUTATION") == "true" && !string.IsNullOrWhiteSpace(context?.ApprovalGrant))
        {
            preview["execution_status"] = "completed";
            preview["approval_required"] = false;
            preview["mutation_performed"] = false;
            preview["note"] = "Approved dataset draft recorded. Raw SQL generation remains deployment-owned.";
        }
        return preview;
    }

    private static Dictionary<string, object?> WritePreview(Dictionary<string, object?> capability, Dictionary<string, object?> plan, string action, Dictionary<string, object?> body, Dictionary<string, object?> metadata) =>
        Result(capability, plan, "prepared", new() { ["approval_required"] = true, ["mutation_performed"] = false, ["superset_action"] = action, ["superset_metadata"] = metadata, ["superset_request"] = new Dictionary<string, object?> { ["operation"] = action, ["body"] = body }, ["note"] = "Prepared a governed Superset analytics request. No Superset mutation was performed." });

    private static string AccessToken()
    {
        var direct = Text(Environment.GetEnvironmentVariable("SUPERSET_ACCESS_TOKEN"));
        if (!string.IsNullOrWhiteSpace(direct)) return direct;
        var username = Text(Environment.GetEnvironmentVariable("SUPERSET_USERNAME"));
        var password = Text(Environment.GetEnvironmentVariable("SUPERSET_PASSWORD"));
        if (string.IsNullOrWhiteSpace(username) || string.IsNullOrWhiteSpace(password)) return "";
        var payload = RequestJson("POST", "/api/v1/security/login", "", new() { ["username"] = username, ["password"] = password, ["provider"] = FirstNonEmpty(Environment.GetEnvironmentVariable("SUPERSET_AUTH_PROVIDER"), "db"), ["refresh"] = true });
        return Text(payload.GetValueOrDefault("access_token"));
    }

    private static Dictionary<string, object?> RequestJson(string method, string path, string token, Dictionary<string, object?>? body)
    {
        using var request = new HttpRequestMessage(new HttpMethod(method), $"{FirstNonEmpty(Environment.GetEnvironmentVariable("SUPERSET_BASE_URL"), "http://127.0.0.1:18088").TrimEnd('/')}{path}");
        request.Headers.Accept.ParseAdd("application/json");
        request.Headers.UserAgent.ParseAdd("anip-superset-fronting-showcase");
        if (!string.IsNullOrWhiteSpace(token)) request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", token);
        if (body is not null) request.Content = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");
        try
        {
            using var response = Http.Send(request);
            var text = response.Content.ReadAsStringAsync().GetAwaiter().GetResult();
            var payload = string.IsNullOrWhiteSpace(text) ? new Dictionary<string, object?>() : JsonSerializer.Deserialize<Dictionary<string, object?>>(text)!;
            if (!response.IsSuccessStatusCode) return new() { ["error"] = "superset_http_error", ["status"] = (int)response.StatusCode, ["detail"] = payload };
            return payload;
        }
        catch (Exception exc)
        {
            return new() { ["error"] = "superset_connection_error", ["detail"] = exc.ToString() };
        }
    }

    private static Dictionary<string, object?> Result(Dictionary<string, object?> capability, Dictionary<string, object?> plan, string status, Dictionary<string, object?> extra)
    {
        var result = new Dictionary<string, object?> { ["execution_status"] = status, ["capability_id"] = capability.GetValueOrDefault("capability_id"), ["selected_backend"] = plan.GetValueOrDefault("selected_binding"), ["semantic_input"] = plan.GetValueOrDefault("semantic_input"), ["backend_input_contract"] = plan.GetValueOrDefault("backend_input_contract") };
        foreach (var item in extra) result[item.Key] = item.Value;
        return result;
    }

    private static Dictionary<string, object?> Restricted(Dictionary<string, object?> capability, Dictionary<string, object?> plan, string reason) =>
        Result(capability, plan, "restricted", new() { ["reason"] = reason });

    private static List<Dictionary<string, object?>> ListResult(Dictionary<string, object?> payload)
    {
        var result = new List<Dictionary<string, object?>>();
        if (payload.GetValueOrDefault("result") is JsonElement element && element.ValueKind == JsonValueKind.Object && element.TryGetProperty("data", out var data) && data.ValueKind == JsonValueKind.Array)
        {
            foreach (var item in data.EnumerateArray())
            {
                result.Add(JsonSerializer.Deserialize<Dictionary<string, object?>>(item.GetRawText())!);
            }
        }
        return result;
    }

    private static string Get(Dictionary<string, object?> payload, string key) => payload.GetValueOrDefault(key) is JsonElement element ? (element.ValueKind == JsonValueKind.String ? element.GetString() ?? "" : element.ToString()) : Text(payload.GetValueOrDefault(key));
    private static bool ScopeAllowed(string scope) => !CsvEnv("ANIP_SUPERSET_BLOCKED_WORKSPACES").Contains(scope.ToLowerInvariant()) && (CsvEnv("ANIP_SUPERSET_ALLOWED_WORKSPACES").Count == 0 || CsvEnv("ANIP_SUPERSET_ALLOWED_WORKSPACES").Contains(scope.ToLowerInvariant()));
    private static bool DatasetAllowed(string datasetRef) => CsvEnv("ANIP_SUPERSET_ALLOWED_DATASETS").Count == 0 || CsvEnv("ANIP_SUPERSET_ALLOWED_DATASETS").Contains(datasetRef.ToLowerInvariant());
    private static int BoundedLimit(object? value, int defaultValue, int maximum) => int.TryParse(Text(value), out var parsed) ? Math.Max(1, Math.Min(parsed, maximum)) : defaultValue;
    private static string FirstNonEmpty(params object?[] values) => values.Select(Text).FirstOrDefault(value => !string.IsNullOrWhiteSpace(value)) ?? "";
    private static string Text(object? value) => value is null ? "" : value.ToString()!.Trim();
    private static List<string> CsvEnv(string name) => (Environment.GetEnvironmentVariable(name) ?? "").Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries).Select(item => item.ToLowerInvariant()).Distinct().ToList();
}
